core-rust/state-manager/src/staging/cache.rs

```diff
@@ -170,41 +243,108 @@ impl<'s, S: ReadableSubstateStore> ReadableSubstateStore for StagedSubstateStore
     }
 }
 
-impl Delta for TransactionReceipt {
+impl<'s, S: RootStore> ReadableTreeStore for StagedStore<'s, S> {
+    fn get_node(&self, key: &NodeKey) -> Option<TreeNode> {
+        self.overlay
+            .hash_tree_nodes
+            .get(key)
+            .cloned()
+            .or_else(|| ReadableTreeStore::get_node(self.root, key))
+    }
+}
+
+impl ProcessedResult {
+    pub fn new(
+        state_hash: Hash,
+        receipt: TransactionReceipt,
+        hash_tree_diff: HashTreeDiff,
+    ) -> Self {
+        let empty_state_diff = if let TransactionResult::Commit(_) = &receipt.result {
+            None
+        } else {
+            Some(StateDiff::new())
+        };
+        Self {
+            state_hash,
+            receipt,
+            empty_state_diff,
+            hash_tree_diff,
+        }
+    }
+    pub fn receipt(&self) -> &TransactionReceipt {
+        &self.receipt
+    }
+
+    pub fn state_diff(&self) -> &StateDiff {
```

Comment:
This could return `Option<&StateDiff>` to save storing the `empty_state_diff` - give me a shout if you'd like to pair on this 👍